File: src/domain/intake.ts
export type SourceKind = "file" | "folder";
export type ProjectKind = "git" | "folder" | "notes";

export interface IntakeSource {
  path: string;
  kind: SourceKind;
  readOnly: boolean;
}

export interface ProjectSummary {
  projectId: string;
  root: string;
  kind: ProjectKind;
  languages: string[];
  buildSystems: string[];
  testCommands: string[];
  warnings: string[];
}

export interface IntakeManifest {
  schemaVersion: number;
  intakeId: string;
  sources: IntakeSource[];
  projects: ProjectSummary[];
  inventoryArtifact: string;
}

export interface InventorySummary {
  fileCount: number;
  dirCount: number;
  totalBytes: number;
  ignoredCount: number;
  truncated: boolean;
  truncationReason?: string;
  estimatedCopyBytes: number;
}

export interface TrustState {
  confirmed: boolean;
  acknowledgedUntrusted: boolean;
  acknowledgedExecutionRisk: boolean;
}

export interface SecretRiskFinding {
  relativePath: string;
  patternId: string;
  matchHash: string;
  matchByteLen: number;
}

export interface CursorProbeStub {
  status: string;
  message: string;
  executable?: string;
  version?: string;
  auth?: string;
  modelCount?: number;
  hasNoninteractiveApproval?: boolean;
}

export interface IntakeLimits {
  maxFiles: number;
  maxTotalBytes: number;
  maxFileBytes: number;
  maxSecretScanBytes: number;
  maxDepth: number;
}

export interface PreflightReport {
  schemaVersion: number;
  manifest: IntakeManifest;
  inventory: InventorySummary;
  warnings: string[];
  blockers: string[];
  secretRisks: SecretRiskFinding[];
  escapeAttempts: string[];
  trust: TrustState;
  cursor: CursorProbeStub;
  canStart: boolean;
  readRoots: string[];
  writeRootsPreview: string[];
  limits: IntakeLimits;
  untrustedContentNotice: string;
}

export function canStartImplementation(report: PreflightReport | null): boolean {
  return Boolean(report?.canStart);
}

File: crates/tiamat-contracts/src/intake.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct IntakeSource {
    pub path: String,
    pub kind: SourceKind,
    pub read_only: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum SourceKind {
    File,
    Folder,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ProjectKind {
    Git,
    Folder,
    Notes,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct ProjectSummary {
    pub project_id: String,
    pub root: String,
    pub kind: ProjectKind,
    pub languages: Vec<String>,
    pub build_systems: Vec<String>,
    pub test_commands: Vec<String>,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct IntakeManifest {
    pub schema_version: u32,
    pub intake_id: Uuid,
    pub sources: Vec<IntakeSource>,
    pub projects: Vec<ProjectSummary>,
    pub inventory_artifact: String,
}

impl IntakeManifest {
    pub fn validate_schema_version(&self) -> Result<(), crate::validation::ValidationError> {
        if self.schema_version != crate::CURRENT_SCHEMA_VERSION {
            return Err(
                crate::validation::ValidationError::IncompatibleSchemaVersion {
                    expected: crate::CURRENT_SCHEMA_VERSION,
                    found: self.schema_version,
                },
            );
        }
        Ok(())
    }
}
